common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -428,121 +450,138 @@ public boolean close(T trackedObject) {
 
         @Override
         public String toString() {
-            if (head == null) {
+            Record oldHead;
+            do {
+                oldHead = headUpdater.get(this);
+            } while (!headUpdater.compareAndSet(this, oldHead, null));
+            if (oldHead == null) {
+                // Already closed
                 return EMPTY_STRING;
             }
 
-            final String creationRecord;
-            final String[] array;
-            int idx = 0;
-            String last = null;
-            final int dropped;
-
-            synchronized (head) {
-                if (tail == null) {
-                    // Already closed
-                    return EMPTY_STRING;
-                }
-                dropped = droppedRecords;
-                creationRecord = head.toString();
-                array = new String[numRecords];
-                Record record = head.next;
-                while (record != null) {
-                    String recordStr = record.toString();
-                    if (last == null || !last.equals(recordStr)) {
-                        array[idx ++] = recordStr;
-                        last = recordStr;
+            final int dropped = droppedRecordsUpdater.get(this);
+            int duped = 0;
+
+            int present = oldHead.pos + 1;
+            // Guess about 2 kilobytes per stack trace
+            StringBuilder buf = new StringBuilder(present * 2048).append(NEWLINE);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) StringBuffer (or StringBuilder).append is called consecutively without reusing the target variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AConsecutiveAppendsShouldReuse)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) StringBuffer constructor is initialized with size 16, but has at least 23 characters appended. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AInsufficientStringBufferDeclaration)